File: netlogo-headless/src/main/prim/etc/_wait.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc

import org.nlogo.core.Let
import org.nlogo.nvm.{ AssemblerAssistant, Command, Context, CustomAssembled, MutableLong }

class _wait extends Command with CustomAssembled {

  private val let = Let("~WAITCOUNTER")

  override def perform(context: Context): Unit = {
    val now = System.nanoTime()
    val duration = (argEvalDoubleValue(context, 0) * 1000000000.0).toLong
    context.activation.binding.let(let, new MutableLong(now + duration))
    context.ip = next
  }

  override def assemble(a: AssemblerAssistant): Unit = {
    a.add(this)
    a.add(new _waitinternal(let))
  }

}

File: netlogo-gui/src/main/prim/etc/_wait.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc;

import org.nlogo.core.Let;
import org.nlogo.api.LogoException;
import org.nlogo.core.Syntax;
import org.nlogo.nvm.Command;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.MutableLong;

public final strictfp class _wait
    extends Command
    implements org.nlogo.nvm.CustomAssembled {

  private final Let let = new Let("~WAITCOUNTER");

  @Override
  public void perform(final Context context)
      throws LogoException {
    long targetTime = System.nanoTime();
    targetTime += argEvalDoubleValue(context, 0) * 1000000000;
    context.activation.binding.let(let, new MutableLong(targetTime));
    context.ip = next;
  }

  public void assemble(org.nlogo.nvm.AssemblerAssistant a) {
    a.add(this);
    a.add(new _waitinternal(let));
  }
}
